fs/littlefs/src/mynewt_glue.c

```diff
@@ -0,0 +1,839 @@
+           uint32_t *out_len)
+{
+    lfs_file_t *file;
+    lfs_t *lfs;
+    int32_t size;
+
+    if (!fs_file || !out_data || !out_len) {
+        return FS_EINVAL;
+    }
+
+    if (!len) {
+        *out_len = 0;
+        return FS_EOK;
+    }
+
+    file = ((struct littlefs_file *) fs_file)->file;
+    lfs = ((struct littlefs_file *) fs_file)->lfs;
+
+    littlefs_lock();
+    size = lfs_file_read(lfs, file, out_data, *out_len);
+    littlefs_unlock();
+    if (size < 0) {
+        return littlefs_to_vfs_error((int)size);
+    }
+
+    *out_len = (uint32_t)size;
+    return FS_EOK;
+}
+
+static int
+littlefs_write(struct fs_file *fs_file, const void *data, int len)
+{
+    lfs_file_t *file;
+    lfs_t *lfs;
+    int32_t size;
+
+    if (!fs_file || !data) {
+        return FS_EINVAL;
+    }
+
+    if (!len) {
+        return FS_EOK;
+    }
+
+    file = ((struct littlefs_file *) fs_file)->file;
+    lfs = ((struct littlefs_file *) fs_file)->lfs;
+
+    littlefs_lock();
+    size = lfs_file_write(lfs, file, data, len);
+    littlefs_unlock();
+    if (size < 0) {
+        return littlefs_to_vfs_error((int)size);
+    }
+
+    if (size != len) {
+        return FS_EFULL;
+    }
+
+    return FS_EOK;
+}
+
+static int
+littlefs_flush(struct fs_file *fs_file)
+{
+    (void)fs_flush;
+    return FS_EOK;
+}
+
+static int
+littlefs_unlink(const char *path)
+{
+    int rc;
+
+    if (!path) {
+        return FS_EINVAL;
+    }
+
+    littlefs_lock();
+    rc= lfs_remove(g_lfs, path);
+    littlefs_unlock();
+
+    return littlefs_to_vfs_error(rc);
+}
+
+static int
+littlefs_rename(const char *from, const char *to)
+{
+    int rc;
+
+    if (!from || !to) {
+        return FS_EINVAL;
+    }
+
+    littlefs_lock();
+    rc = lfs_rename(g_lfs, from, to);
+    littlefs_unlock();
+
+    return littlefs_to_vfs_error(rc);
+}
+
+static int
+littlefs_mkdir(const char *path)
+{
+    int rc;
+
+    if (!path) {
+        return FS_EINVAL;
+    }
+
+    littlefs_lock();
+    rc = lfs_mkdir(g_lfs, path);
+    littlefs_unlock();
+
+    return littlefs_to_vfs_error(rc);
+}
+
+static int
+littlefs_opendir(const char *path, struct fs_dir **out_fs_dir)
+{
+    lfs_dir_t *out_dir;
+    struct littlefs_dir *dir;
+    int rc;
+
+    if (!path || !out_fs_dir) {
+        return FS_EINVAL;
+    }
+
+    out_dir = NULL;
+
+    dir = malloc(sizeof(struct littlefs_dir));
+    if (!dir) {
+        rc = FS_ENOMEM;
+        goto out;
+    }
+
+    out_dir = malloc(sizeof(lfs_dir_t));
+    if (!out_dir) {
+        rc = FS_ENOMEM;
+        goto out;
+    }
+
+    littlefs_lock();
+    rc = lfs_dir_open(g_lfs, out_dir, path);
+    littlefs_unlock();
+    if (rc < 0) {
+        rc = littlefs_to_vfs_error(rc);
+        goto out;
+    }
+
+    dir->dir = out_dir;
+    dir->cur_dirent = NULL;
+    dir->fops = &littlefs_ops;
+    dir->lfs = g_lfs;
+    *out_fs_dir = (struct fs_dir *)dir;
+    rc = FS_EOK;
+
+out:
+    if (rc != FS_EOK) {
+        if (dir) free(dir);
```

Comment:
style
also doesn't free() already check for null pointers?

Reply:
Fixed.